### DISCUSSION TITLE
Creating repository rulesets using the REST API

### DISCUSSION THREAD

**Discussion by philnewm** [community/community#139808] [Apps, API and Webhooks] [answered]
### Select Topic Area

Question

### Body

I'm currently trying to implement the creation of repository rulesets using the GitHub REST API.

This is intended in the context of a github actions workflow which is supposed to automatically create a repository for testing purposes.
I would like to use the gh tool in this case and read through the [docs](https://docs.github.com/en/rest/repos/rules?apiVersion=2022-11-28#create-a-repository-ruleset)
If a Question like this already exist, please point me to it since I couldn't find one.

I'm also open to other approaches in case there are any.

So far I implemented one rulesets which works just fine but it's also pretty limited.
This second ruleset includes way more details and causes me errors.
```yaml
- name: Apply ruleset for 'develop' branch
  run: |
    gh api \
      --method POST \
      -H "Accept: application/vnd.github+json" \
      -H "X-GitHub-Api-Version: 2022-11-28" \
      /repos/<owner>/<repo-name>/rulesets \
      -f "name=Protect develop" \
      -f "target=branch" \
      -f "enforcement=active" \
      -f "conditions[ref_name][include][]=refs/heads/develop" \
      -f "conditions[ref_name][exclude][]" \
      -f "rules[][type]=deletion" \
      -f "rules[][type]=non_fast_forward" \
      -F "rules[][type]=required_status_checks" \
      -F "rules[][parameters][required_status_checks][context]=linting" \
      -F "rules[][parameters][required_status_checks][integration_id]=15368" \
      -f "rules[][parameters][strict_required_status_checks_policy]=true" \
      -f "rules[][parameters][do_not_enforce_on_create]=false" \
      -f "rules[][type]=pull_request" \
      -F "rules[][parameters][required_approving_review_count]=1" \
      -f "rules[][parameters][dismiss_stale_reviews_on_push]=false" \
      -f "rules[][parameters][require_code_owner_review]=true" \
      -f "rules[][parameters][require_last_push_approval]=false" \
      -f "rules[][parameters][required_review_thread_resolution]=false" \
      -F "bypass_actors[][actor_id]=<id>" \
      -f "bypass_actors[][actor_type]=RepositoryRole" \
      -f "bypass_actors[][bypass_mode]=always" \
      -F "bypass_actors[][actor_id]=<id>" \
      -f "bypass_actors[][actor_type]=Team" \
      -f "bypass_actors[][bypass_mode]=always" \
      -F "bypass_actors[][actor_id]=1" \
      -f "bypass_actors[][actor_type]=OrganizationAdmin" \
      -f "bypass_actors[][bypass_mode]=always"
```

I keep getting this error output
```
{
  "message": "Invalid request.\n
\n
Invalid property /bypass_actors/3: object is missing required keys: actor_type, bypass_mode.\n
Invalid property /bypass_actors/4: object is missing required keys: actor_type, bypass_mode.\n
Invalid property /rules/2: data matches no possible input. See `documentation_url`.\n
Invalid property /rules/3: data matches no possible input. See `documentation_url`.\n
Invalid property /rules/4: data matches no possible input. See `documentation_url`.\n
Invalid property /rules/5: data matches no possible input. See `documentation_url`.\n
Invalid property /rules/6: data matches no possible input. See `documentation_url`.\n
Invalid property /rules/7: data matches no possible input. See `documentation_url`.\n
Invalid property /rules/8: data matches no possible input. See `documentation_url`.\n
Invalid property /rules/9: data matches no possible input. See `documentation_url`.",
  "documentation_url": "https://docs.github.com/rest/repos/rules#create-a-repository-ruleset",
  "status": "422"
}
gh: Invalid request.

Invalid property /bypass_actors/3: object is missing required keys: actor_type, bypass_mode.
Invalid property /bypass_actors/4: object is missing required keys: actor_type, bypass_mode.
Invalid property /rules/2: data matches no possible input. See `documentation_url`.
Invalid property /rules/3: data matches no possible input. See `documentation_url`.
Invalid property /rules/4: data matches no possible input. See `documentation_url`.
Invalid property /rules/5: data matches no possible input. See `documentation_url`.
Invalid property /rules/6: data matches no possible input. See `documentation_url`.
Invalid property /rules/7: data matches no possible input. See `documentation_url`.
Invalid property /rules/8: data matches no possible input. See `documentation_url`.
Invalid property /rules/9: data matches no possible input. See `documentation_url`. (HTTP 422)
```

It's in some way related to the syntax I used for rule and bypass_actor parameters.
From the examples in the documentation I could not really tell how this is intended to be used.
I assume I'm using wrong syntax in cases of nesteed object arrays like for `required_status_checks` but not sure.

Would really appreciate some help with this one.

Cheers

**Comment by philnewm** [comment]
Bump - still need help on this one.

**Comment by ThisIs-Developer** [comment]
hey @philnewm,

it looks like the error is related to the structure of the **rules** and **bypass_actors** arrays. a **422** status usually means there's an issue with the data format or missing keys in your request.

1. **bypass_actors**: 
   make sure each **bypass_actor** entry has both `actor_type` and `bypass_mode` fields filled. from your example, it seems some of these entries are missing or incomplete.
   
   - double-check entries like `bypass_actors[][actor_id]=1` to ensure the **actor_type** and **bypass_mode** are specified.

2. **rules**: 
   the **rules** part of the request might have some incorrect nesting. verify each rule's structure against the [github api documentation](https://docs.github.com/rest/repos/rules#create-a-repository-ruleset). focus on the **required_status_checks** section

3. use the api in debug mode or look for more detailed logs to trace where the issue is in the object structure.

hope this helps!

---
if this works for you, feel free to mark it as the answer! 😪

**Reply by philnewm** [reply]
Thansk for your answer, I'll try this today and report back.

**Reply by ThisIs-Developer** [reply]
Ok

**Comment by philnewm** [comment, marked as answer]
I made it work now. But I went down a slightly different road.

So an approach that did work is just using a json file - which I exported from the GitHub ruleset UI
After all the steps in my github workflow now look like this and are fully functional
```yaml
- name: Apply ruleset for '${{ env.MAIN_BRANCH }}'
  run: |
    curl -O https://raw.githubusercontent.com/${{ github.repository }}/${{ env.MAIN_BRANCH }}/branch_rules/protect_main.json
    gh api \
      --method POST \
      -H "Accept: application/vnd.github+json" \
      -H "X-GitHub-Api-Version: 2022-11-28" \
      /repos/${{ env.TARGET_REPO }}/rulesets \
      --input protect_main.json

- name: Apply ruleset for 'develop'
  run: |
    curl -O https://raw.githubusercontent.com/${{ github.repository }}/${{ env.MAIN_BRANCH }}/branch_rules/protect_develop.json
    gh api \
      --method POST \
      -H "Accept: application/vnd.github+json" \
      -H "X-GitHub-Api-Version: 2022-11-28" \
      /repos/${{ env.TARGET_REPO }}/rulesets \
      --input protect_develop.json
```

@ThisIs-Developer thanks for your answer - I suppose your pointing towards the docs helped me in the end.
